### PR TITLE
Adding System.IO.Pipelines to replace Channels and Streams (replaces #949)

### LINKED ISSUE #970: Proposal: Full async channel interface (IModel)
I've drafted a proposal for a new IModel interface that I tried to keep as slim as possible while changing everything relevant to Async.
**This proposal is in no way finalized and should serve as a discussion base. (I'll update the proposal accordingly)**
Please also verify the naming of the members and methods, I think with the breaking change that we do anyway, we should make sure the new names are exactly what we want.

A few things to highlight:
- I renamed the interface to IChannel to match other languages.
- I changed the naming generally to reflect what it is doing away from the protocol names used previously (I think they're confusing for a user without the background of the protocol).
- I named the methods in a style where it begins with a verb and all async methods have the Async Postfix. Events in past tense.
- All Async methods return ValueTask or ValueTask<T>
- All passive methods were replaced by an additional parameter "throwOnMismatch"
- All noWait methods were replaced by an additional parameter "waitForConfirmation"
- A few methods were removed (comments suggested deprecation or I felt they are not useful)
- I've added several todos at places where things are still unclear / need discussion
- I've tried to group them for better visibility.

Prerequisites:
- Used ValueTuple => Not possible for net461
- Used IAsyncDisposable => Not possible for net461, Reference Microsoft.Bcl.AsyncInterfaces or target .net core 3.1
- Used nullable reference types => Not possible for net461, [Workaround possible for .netstandard 2.0](https://www.meziantou.net/how-to-use-nullable-reference-types-in-dotnet-standard-2-0-and-dotnet-.htm)

[As a reference, here's the current interface.](https://github.com/rabbitmq/rabbitmq-dotnet-client/blob/master/projects/RabbitMQ.Client/client/api/IModel.cs)
```C#
    public interface IChannel : IDisposable, IAsyncDisposable
    {
        int ChannelNumber { get; }
        bool IsOpen { get; }
        TimeSpan ContinuationTimeout { get; set; }

        /****************************
         * Various methods
         ****************************/
        ValueTask SetQosAsync(uint prefetchSize, ushort prefetchCount, bool global);
        // exception, context-detail
        event Action<Exception, Dictionary<string, object>?>? UnhandledExceptionOccurred;
        // active, TODO ask is this still needed?
        event Action<bool>? FlowControlChanged;

        /****************************
         * Message retrieval methods
         ****************************/
        ValueTask<string> ActivateConsumerAsync(IBasicConsumer consumer, string queue, bool autoAck, string consumerTag = "", bool noLocal = false, bool exclusive = false, IDictionary<string, object>? arguments = null);
        ValueTask CancelConsumerAsync(string consumerTag, bool waitForConfirmation = true);
        // TODO return value depends on how we represent messages
        ValueTask<BasicGetResult?> RetrieveSingleMessageAsync(string queue, bool autoAck);
        ValueTask AckMessageAsync(ulong deliveryTag, bool multiple);
        ValueTask NackMessageAsync(ulong deliveryTag, bool multiple, bool requeue);
        // TODO ask is this still needed?
        ValueTask RejectMessageAsync(ulong deliveryTag, bool requeue);

        /****************************
         * Message publication methods
         ****************************/
        event MessageDeliveryFailedDelegate? MessageDeliveryFailed;
        ValueTask PublishMessageAsync(string exchange, string routingKey, IBasicProperties? basicProperties, ReadOnlyMemory<byte> body, bool mandatory = false);
        // Depends on #972 
        ValueTask PublishMessageAsync(ReadOnlyMemory<byte> exchange, ReadOnlyMemory<byte> routingKey, IBasicProperties? basicProperties, ReadOnlyMemory<byte> body, bool mandatory = false);
        ValueTask ActivatePublishAcksAsync();
        // delivery tag, multiple, isAck
        event Action<ulong, bool, bool>? PublishAckReceived;
        event Action<ulong>? NewPublishTagUsed;
        ValueTask<ulong> PublishAckableMessageAsync(string exchange, string routingKey, IBasicProperties? basicProperties, ReadOnlyMemory<byte> body, bool mandatory = false);

        /****************************
         * Close methods
         ****************************/
        // CloseReason
        event Action<ShutdownEventArgs>? Shutdown;
        ShutdownEventArgs? CloseReason { get; }
        ValueTask AbortAsync(ushort replyCode, string replyText);
        ValueTask CloseAsync(ushort replyCode, string replyText);

        /****************************
         * Exchange methods
         ****************************/
        ValueTask DeclareExchangeAsync(string exchange, string type, bool durable, bool autoDelete, bool throwOnMismatch = true, IDictionary<string, object>? arguments = null, bool waitForConfirmation = true);
        ValueTask DeleteExchangeAsync(string exchange, bool ifUnused = false, bool waitForConfirmation = true);
        ValueTask BindExchangeAsync(string destination, string source, string routingKey, IDictionary<string, object>? arguments = null, bool waitForConfirmation = true);
        ValueTask UnbindExchangeAsync(string destination, string source, string routingKey, IDictionary<string, object>? arguments = null, bool waitForConfirmation = true);

        /****************************
         * Queue methods
         ****************************/
        ValueTask<(string QueueName, uint MessageCount, uint ConsumerCount)> DeclareQueueAsync(string queue, bool durable, bool exclusive, bool autoDelete, bool throwOnMismatch = true, IDictionary<string, object>? arguments = null);
        ValueTask DeclareQueueWithoutConfirmationAsync(string queue, bool durable, bool exclusive, bool autoDelete, IDictionary<string, object>? arguments = null);
        ValueTask<uint> DeleteQueueAsync(string queue, bool ifUnused = false, bool ifEmpty = false);
        ValueTask DeleteQueueWithoutConfirmationAsync(string queue, bool ifUnused = false, bool ifEmpty = false);
        ValueTask BindQueueAsync(string queue, string exchange, string routingKey, IDictionary<string, object>? arguments = null, bool waitForConfirmation = true);
        ValueTask UnbindQueueAsync(string queue, string exchange, string routingKey, IDictionary<string, object>? arguments = null);
        ValueTask<uint> PurgeQueueAsync(string queue);

        /****************************
         * Transaction methods
         ****************************/
        ValueTask ActivateTransactionsAsync();
        ValueTask CommitTransactionAsync();
        ValueTask RollbackTransactionAsync();
    }

    // Delegate definitions due to "in" modifier
    public delegate void MessageDeliveryFailedDelegate(in Message message, string replyText, ushort replyCode);
 
    // Auxiliary types
    public readonly struct Message
    {
        public readonly string Exchange;
        public readonly string RoutingKey;
        public readonly IBasicProperties Properties;
        public readonly ReadOnlyMemory<byte> Body;

        public Message(string exchange, string routingKey, IBasicProperties properties, ReadOnlyMemory<byte> body)
        {
            Exchange = exchange;
            RoutingKey = routingKey;
            Properties = properties;
            Body = body;
        }
    }

    // Extension methods
    public static class ChannelExtensions
    {
        // Queue methods
        ValueTask<(string QueueName, uint MessageCount, uint ConsumerCount)> DeclareQueueAsync(this Channel channel, string queue = "", bool durable = false, bool exclusive = true, bool autoDelete = true, bool throwOnMismatch = true, IDictionary<string, object>? arguments = null);
        ValueTask<uint> GetQueueMessageCountAsync(this Channel channel, string queue);
        ValueTask<uint> GetQueueConsumerCountAsync(this Channel channel, string queue);

        // Close methods
        ValueTask CloseAsync(this Channel channel);
        ValueTask AbortAsync(this Channel channel);
    }
```

The WaitForConfirms methods I plan to replace with a new class that can be wrapped around IChannel, so customer can decide if they want to use our basic implementation or build something on their own. 
```C#
    public sealed class PublishAckAwaiter : IValueTaskSource<bool>
    {
        public PublishAckAwaiter(IChannel channel) { }

        ValueTask<bool> WaitForAckAsync(ulong publishTag, CancellationToken cancellation) { }
        ValueTask<bool> WaitForAckOrCloseAsync(ulong publishTag, CancellationToken cancellation) { }
        ValueTask<bool> WaitForAllAcksAsync(CancellationToken cancellation) {}
        ValueTask<bool> WaitForAllAcksOrCloseAsync(CancellationToken cancellation) {}
    }
```

FYI removed methods:
```C#
    public interface IModel : IDisposable
    {
        // Does anyone really use it? Removed due to not known by me.
        IBasicConsumer DefaultConsumer { get; set; }
        // not needed if we have IsOpen & ShutdownReason
        bool IsClosed { get; }
        // replaced by return value on publish method
        ulong NextPublishSeqNo { get; }

        // deprecated according to comments
        event EventHandler<EventArgs> BasicRecoverOk;
        void BasicRecover(bool requeue);
        void BasicRecoverAsync(bool requeue);

        // removed from interface, can be delivered as extension methods
        void Abort();
        void Close();

        // "new BasicProperty()" / "new BasicPublishBatch()" => no need to hide them behind these methods
        IBasicPublishBatch CreateBasicPublishBatch();
        IBasicPublishBatch CreateBasicPublishBatch(int sizeHint);
        IBasicProperties CreateBasicProperties();
    }
```

### PR DESCRIPTION
**Pull request by stebet**
## Proposed Changes

This PR adds [System.IO.Pipelines](https://devblogs.microsoft.com/dotnet/system-io-pipelines-high-performance-io-in-net/) to replace using Channels and Streams to receive and send data to/from the client.

* Removes Channels/Streams and replaces them with PipeReader/PipeWriter, which are now handled by the Connection instead of SocketFrameHandler
  * Simplifies the SocketFrameHandler a lot! Gets rid of extra Read/Write tasks which are now managed by the Pipelines.
  * Does require sync-over-async ugliness to Flush the Pipe when publishes manage to fill the Pipe buffer faster than data gets sent. That is however rare under normal circumstances.
* Prepares the client to be able to support fully asynchronous operations more easily.
  * This does create a little bit of CPU/Allocation overhead for the async operations under mass publish scenarios (benchmarks incoming), but we would always have run into those issues when making the operations Async anyway or when fixing the potential OOM issues like in #1106 due to having to create "fake" tasks to block while waiting for the buffers to drop below the threshold levels. This overhead should however be negligible under normal scenarios and should actually be a lot better off under fully async scenarios since we are less likely to block threads and create ThreadPool starvation scenarios.
* Should solve several problems similar to #1106, which stem from the current client being able publish messages faster than the network connection/RabbitMQ server can handle them for async scenarios.
* Lays groundwork to fix #970, #966 and #843

## Types of Changes

What types of changes does your code introduce to this project?
_Put an `x` in the boxes that apply_

- [X] Bug fix (non-breaking change which fixes issue #1106)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause an observable behavior change in existing systems)
- [ ] Documentation improvements (corrections, new content, etc)
- [ ] Cosmetic change (whitespace, formatting, etc)

## Checklist

- [X] I have read the `CONTRIBUTING.md` document
- [X] I have signed the CA (see https://cla.pivotal.io/sign/rabbitmq)
- [X] All tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in related repositories

### Reasoning
I see Pipelines being a perfect fit here. They make processing buffers a lot easier, reduce overhead when deserializing data and make it a lot easier to implement fully async operations. That's also why they are used for the same purpose in [Kestrel](https://github.com/dotnet/aspnetcore/blob/main/src/Servers/Kestrel/Core/src/Internal/Http/Http1Connection.cs) (the ASP.NET Core server) Kestrel and [StackExchange.Redis](https://github.com/StackExchange/StackExchange.Redis/blob/main/src/StackExchange.Redis/PhysicalConnection.cs).

I also created as-of-yet unused [WriteAsync](https://github.com/rabbitmq/rabbitmq-dotnet-client/compare/main...stebet:pipelines?expand=1#diff-09ed630152a341b50003a5c1da0fb23eecb62eec4df30df695f084bd14f52760R410) methods for that purpose, which would make it easy to implement fully Async publish methods for example.

### Benchmark results (to be added)

### CODE REVIEW

**Comment by stebet** [comment]
Still ironing out a few things that only seem to manifest in the CI tests (tests all run fine locally), but think they have to do with the fact that synchronous publish now potentially blocks where before it had an unlimited buffer with the UnboundedChannel to write into, so WaitForConfirm tests behave slightly differently with Pipelines.

File: projects/RabbitMQ.Client/client/impl/Connection.cs [hunk 1]
```diff
@@ -394,9 +397,93 @@ internal void OnCallbackException(CallbackExceptionEventArgs args)
...
         }
 
-        internal void Write(ReadOnlyMemory<byte> memory)
+        internal void Write(ReadOnlySpan<byte> payload)
+        {
+            if (!_frameHandler.IsClosed)
+            {
+                _writeSemaphore.Wait();
+                WriteSpan(payload);
+                _frameHandler.FrameWriter.FlushAsync().AsTask().Wait(); // Sync-over-async, not great, not terrible. All we can do for the sync path.
```
Comment (danielmarbach) [inline projects/RabbitMQ.Client/client/impl/Connection.cs:406]: I have not yet used the pipeline stuff very intensively. I was wondering if we could use `AsStream` and then use the synchronous stream API here and in the other sync parts?

File: projects/RabbitMQ.Client/client/impl/Connection.Receive.cs [hunk 2]
```diff
@@ -149,7 +199,7 @@ private void HardProtocolExceptionHandler(HardProtocolException hpe)
                 {
                     var cmd = new ConnectionClose(hpe.ShutdownReason.ReplyCode, hpe.ShutdownReason.ReplyText, 0, 0);
                     _session0.Transmit(ref cmd);
-                    ClosingLoop();
+                    await ClosingLoop();
```
Comment (danielmarbach) [inline projects/RabbitMQ.Client/client/impl/Connection.Receive.cs:202]: ConfigureAwait

File: projects/RabbitMQ.Client/client/impl/Connection.cs [hunk 3]
```diff
@@ -394,9 +397,93 @@ internal void OnCallbackException(CallbackExceptionEventArgs args)
...
+            if (!_frameHandler.IsClosed)
+            {
+                if (!_writeSemaphore.Wait(0))
+                {
+                    await _writeSemaphore.WaitAsync().ConfigureAwait(false);
+                }
+
+                WriteSpan(payload.Span);
+                await _frameHandler.FrameWriter.FlushAsync().ConfigureAwait(false);
+                _writeSemaphore.Release();
```
Comment (danielmarbach) [inline projects/RabbitMQ.Client/client/impl/Connection.cs:422]: Are you deliberately not doing this in a finally block?

**Review by danielmarbach — COMMENTED** [review]
Sorry only had like five minutes today. Hopefully I can spend some more time another day

**Comment by stebet** [comment]
Yeah I'm working on cleaning it up a bit. Found some test issues and I'm working on a PR to clean those up first to reduce the noise.

File: projects/RabbitMQ.Client/client/impl/Connection.cs (hunk 1 above)
Comment (stebet) [reply projects/RabbitMQ.Client/client/impl/Connection.cs:406]: That actually makes some sense. I'll look into that.

File: projects/RabbitMQ.Client/client/impl/Connection.cs (hunk 3 above)
Comment (stebet) [reply projects/RabbitMQ.Client/client/impl/Connection.cs:422]: Nope. I'll fix that

**Comment by michaelklishin** [comment]
Sorry to drag you into resolving some non-trivial conflicts @stebet. We decided to go ahead
and merge https://github.com/rabbitmq/rabbitmq-dotnet-client/pull/1202 as it felt like the
right thing to do since we require `net6.0` now for 7.0. This obviously affects all pull requests in flight.

**Comment by michaelklishin** [comment]
I reverted #1202 because it makes back and forward-porting impossible unless all maintained branches adopt file-scoped namespaces.

**Comment by stebet** [comment]
No probs. I'm a bit busy with work and RL these days. Hoping to submit another PR first that cleans up the unit tests a lot and hopefully makes the more reliable and then clean this PR up after that.

**Comment by lukebakken** [comment]
@stebet I've actually got some time to dedicate to releasing version 7.0 of this library. I'll get the most recent conflicts fixed in this branch, as well as address the most recent review comments. We could discuss the remainder of work here and I could take over. Let me know what you think.

**Comment by lukebakken** [comment]
Hi @stebet, I'm wondering if you have time to read [my comment](https://github.com/rabbitmq/rabbitmq-dotnet-client/pull/1199#issuecomment-1156478436) and respond. @bollhals [asked about the status of version 7.0](https://github.com/rabbitmq/rabbitmq-dotnet-client/pull/1027#issuecomment-1254981765) and one major task is completing this PR.

As I said, I'm happy to take over work as long as your `piplines` branch is up-to-date in your fork. Thanks!

**Comment by stebet** [comment]
I need to take a little bit closer look at this since I've been out of the loop a bit. I'll take a look in the next days and give you an update.

**Comment by lukebakken** [comment]
Superseded by #1264

**Pull request closed without merging**